std/experimental/color/rgb.d

```
@@ -0,0 +1,852 @@
+    else
+        enum CanFind = s[0] == c || CanFind!(s[1..$], c);
+}
+template AllIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum AllIn = true;
+    else
+        enum AllIn = CanFind!(s, chars[0]) && AllIn!(s, chars[1..$]);
+}
+template AnyIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum AnyIn = false;
+    else
+        enum AnyIn = CanFind!(s, chars[0]) || AnyIn!(s, chars[1..$]);
+}
+template NoneIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NoneIn = true;
+    else
+        enum NoneIn = !CanFind!(s, chars[0]) && NoneIn!(s, chars[1..$]);
+}
+template NotIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NotIn = char(0);
+    else static if(!CanFind!(s, chars[0]))
+        enum NotIn = chars[0];
+    else
+        enum NotIn = NotIn!(s, chars[1..$]);
+}
+
+unittest
+{
+    static assert(CanFind!("string", 'i'));
+    static assert(!CanFind!("string", 'x'));
+    static assert(AllIn!("string", "sgi"));
+    static assert(!AllIn!("string", "sgix"));
+    static assert(AnyIn!("string", "sx"));
+    static assert(!AnyIn!("string", "x"));
+    static assert(NoneIn!("string", "xyz"));
+    static assert(!NoneIn!("string", "sxyz"));
+}
+
+
+// shorthand alias for all components
+alias AllComponents = TypeTuple!('l','r','g','b','a');
+
+// build mixin code to perform expresions per-element
+template ComponentExpression(string expression, char component, string op)
+{
+    template BuildExpression(string e, char c, string op)
+    {
+        static if(e.length == 0)
+            enum BuildExpression = "";
+        else static if(e[0] == '_')
+            enum BuildExpression = c ~ BuildExpression!(e[1..$], c, op);
+        else static if(e[0] == '#')
+            enum BuildExpression = op ~ BuildExpression!(e[1..$], c, op);
+        else
+            enum BuildExpression = e[0] ~ BuildExpression!(e[1..$], c, op);
+    }
+    enum ComponentExpression =
```

> I arrived at this in combat of unruly spaghetti. I'm open to suggestions...

Composition, composition, composition :)

> unittests should help here.

Ironically, this is one situation where unit tests will be less helpful, since the problem is that incorrect code will be silently accepted in unexpected situations, which are hard to foresee.